ratmenu.1
.TH RATMENU 1 "Oct 10 2001" "Ratpoison Desktop For X"
.SH NAME
ratmenu \- create a menu to run commands
.SH SYNOPSIS
.B ratmenu
[
.BI \-display " displayname"
] [
.BI \-font " fname"
] [
.BI \-fg " foreground-color"
] [
.BI \-bg " background-color"
] [
.BI \-io " item-offset"
] [
.BI \-label " name"
] [
.BI \-shell " prog"
] [
.BI \-align " {left|center|right}"
] [
.BI \-style " {snazzy|dreary}"
] [
.BI \-persist
] [
.B \-version
] [
.IR menuitem\ command
]
\&...
.SH DESCRIPTION
.I ratmenu
is a simple program that accepts a list of menu item and command
pairs on the command line.
It creates a window that consists of nothing but a menu.
When a particular item is selected, the corresponding command is executed.
.PP
.I ratmenu
does not recognize any usage of the rodent.  It is entirely
keystroke operated.
.PP
Menu items must be followed by a command.  The syntax is based on
that of the
.I dialog
program.
.PP
.I ratmenu
accepts the following command line options.
.RS
.TP
.BI \-display " displayname"
Use the X display
.IR displayname ,
instead of the default display.
.TP
.BI \-font " fname"
(X Resource: font)
Use the font
.IR fname ,
instead of the default font.
.TP
.BI \-label " name"
Change both the window and icon labels of the window to
.IR name .
The default label is the last component of the path used to run
.IR ratmenu ,
typically,
.BR ratmenu .
.TP
.BI \-fg " foreground-color" 
(X Resource: fgcolor)
Set the foreground color to
.IR foreground-color .
By default, the foreground color is black.
.TP
.BI \-bg " background-color" 
(X Resource: bgcolor)
Set the background color to
.IR background-color .
By default, the background color is white.
.TP
.BI \-io " item-offset"
Set the first selected item to
.IR item-offset .
By default, the first item is the top-most one, or item-offset 1.  The next
item down would be at item-offset 2.
.TP
.BI \-style " {snazzy|dreary}" 
(X Resource: style)
The default style is
.IR snazzy ,
where the highlight bar, which shows the currently selected item, remains
stationary while all the menu items are rotated up or down when the cursor
keys are moved.  In
.IR dreary
mode, the highlight bar moves up and down the menu as it does on all conventional
keyboard controlled menus.
.TP
.BI \-align " {left|center|right}" 
(X Resource: align)
Align the text of the menu entries to the left, right, or center.
Defaults to
.IR left .
This is different from
.BR 9menu ,
which defaults to
.IR center .
.TP
.BI \-shell " prog"
Use
.I prog
as the shell to run commands, instead of
.BR /bin/sh .
A popular alternative shell is
.IR rc (1).
If the shell cannot be executed,
.I ratmenu
will
.I silently
fall back to using
.BR /bin/sh .
.TP
.BI \-back " prevmenu"
.I prevmenu
is a command or shell script to run when the back key is pressed, usually a
script that has ratmenu showing the previous menu.  Used to create nested
menus, it gives the user a way to back out and return to the previous menu.
Note that you can use this option for other things too.  The command specified
by the
.B -back
option is executed when the user hits one of the "back" keys.
.TP
.BI \-persist 
(X Resource: persist)
Ordinarily, ratmenu exits if a new window pops up on top of it, or if you
switch to another window.  This option tells ratmenu to hang around, like a
regular window.
.TP
.B \-version
This option prints the version of
.I ratmenu
on the standard output, and then exits with an exit value of zero.
.RE
.SH KEYSTROKES
.PP
The
.BI Up
keystrokes move the selection to the next item up.  The
.BI Down
keystrokes move the selection to the next item down.  When
the selection reaches the top or bottom, it scrolls around
to the other side on pressing of the appropriate keystroke. The
.BI Select
keystrokes execute the command corresponding to the currently
selected menu item, and exit ratmenu.  The
.BI Exit
keystrokes quit ratmenu without doing anything.
.RS
.TP
.BI Up
 'k', Up_arrow, BackSpace, '\-', C\-p
.TP
.BI Down
 'j', Down_arrow, Space, Tab, '+', C\-n, C\-i
.TP
.BI Select
 'l', Right_arrow, Return, C\-f, C\-m, C\-j
.TP
.BI Back
 'h', Left_arrow, C\-b
.TP
.BI Exit
 'q', Escape, C\-g
.RE
.SH CONFIGURATION USING X RESOURCES
.PP
You can configure some items using
.I xrdb
to load appropriate X resources.  There is an example configuration
in the file
.I /usr/share/doc/ratmenu/Xresources.ratmenu
which shows all possible configuration items.  Give it a look; it is
fairly easy to figure out.
.RE
.SH EXAMPLES
.ft B
.nf
ratmenu \-label "Remotes xterm" acme "rsh acme xterm" herman "rsh herman 9term" &
.sp
ratmenu \-label 'X progs' ghostview ghostview xdvi xdvi xeyes xeyes xneko xneko &
.sp
ratmenu \-back ~/bin/mypreviousmenu "X Eyes" xeyes &
.ft
.fi
.SH SEE ALSO
.IR ratpoison (1),
.IR ratpoisonrc (5),
.IR dialog (1)
.IR xrdb (1)
.PP
.IR "Ratpoison Desktop Users Manual" .
.SH AUTHORS
The initial idea for this program was by Arnold Robbins, after having
worked with John Mackin's GWM Blit emulation.
Matty Farrow wrote a version using libXg, from which some ideas were borrowed.
This code was written by David Hogan and Arnold Robbins.
Rich Salz motivated the
.B \-shell
option.
Jonathan Walther modified this code to play nicely with the ratpoison window
manager by removing handling of mouse events and iconification. Zrajm C Akfohg
made many improvements in his ratmen fork, including X resource handling, some
of which were folded back in.


ChangeLog
djw Sat Sep 15 15:17:34 PDT 2001 Version 1.0
* Initial release

djw Sat Sep 15 16:17:34 PDT 2001 Version 1.1
* added the -align option so you can decide how menu
  entries get aligned (left, right or center)

djw Sat Sep 15 17:17:34 PDT 2001 Version 1.2
* swapped the sense of the "j" and "k" keystrokes to
  be more natural
* added "q" keystroke to exit for Gergely
* added Space as keystroke same as Return

djw Sun Sep 16 10:22:45 PDT 2001 Version 1.2
* updated the debian directory to reflect the
  change from 9menu to ratmenu

djw Tue Sep 18 11:29:32 PDT 2001 Version 1.3
* made menu behavior configurable with -style option
* added UpArrow/DownArrow and u/d keystroke support
* added Tab and Backspace keystrokes
* changed functionality of Space keystroke to go to
  next item instead of selecting item
* made RightArrow have same functionality as Enter

djw Mon Sep 24 10:52:42 PDT 2001 Version 1.4
* added the -fg, -bg, and -font options back in
* cleaned up the event handling loop to remove
  the rodent detritus
* fixed the man page; support for a special "exit"
  menu item was removed ages ago and now the manpage
  reflects that
* added an install target to the makefile

djw Wed Oct 10 12:57:39 PDT 2001 Version 1.5
* updated manpage to document keystrokes
* took out 'd' and 'u' keystrokes, added '+' and '-'
* changed syntax so menu items can include the ':' character

djw Thu Oct 11 11:56:17 PDT 2001 Version 1.6
* fixed a non-bug; now uses "ratmenu" resource
  class instead of "9menu".

djw Sat Dec 15 12:45:47 PST 2001 Version 1.7
* minor code cleanup; got rid of extraneous call
  to XGetGeometry() and unneeded assignment statements
* fixed the version string

djw Sun Jun  2 11:19:23 PDT 2002 1.8
* fixed typo in usage()
* fixed ChangeLog entry for version 1.7
* added the -back option for full traversal of menu hierarchies

djw Sun Jun  2 13:46:35 PDT 2002 1.9
* updated the manpage

djw Sun Jun  2 14:16:03 PDT 2002 1.9.1
* updated the manpage again to be clearer
* fixed quoting of the ' character in the manpage

djw Sun Jun  8 16:53:43 PDT 2003 2.0
* added the -persist option

djw Sun Jun  8 23:45:51 PDT 2003 2.1
* moved README.9menu into the LICENSE file

djw Sun Feb 13 10:56:24 PST 2005 2.3.2
* removed unneeded fork() from spawn function
* added fallback to default /bin/sh if user specified
  shell didn't work.

djw Wed Mar 30 23:41:25 PST 2005 2.3.4
* fixed uninitialized variable

djw Wed Mar 30 23:41:25 PST 2005 2.3.5
* fixed an improperly initialized variable


ratmenu.c
/*
 * ratmenu.c
 *
 * This program puts up a window that is just a menu, and executes
 * commands that correspond to the items selected.
 *
 * Initial idea: Arnold Robbins
 * Version using libXg: Matty Farrow (some ideas borrowed)
 * This code by: David Hogan and Arnold Robbins
 *
 * Copyright (c), Arnold Robbins and David Hogan
 *
 * Arnold Robbins
 * arnold@skeeve.atl.ga.us
 * October, 1994
 *
 * Code added to cause pop-up (unIconify) to move menu to mouse.
 * Christopher Platt
 * platt@coos.dartmouth.edu
 * May, 1995
 *
 * Said code moved to -teleport option, and -warp option added.
 * Arnold Robbins
 * June, 1995
 *
 * Code added to allow -fg and -bg colors.
 * John M. O'Donnell
 * odonnell@stpaul.lampf.lanl.gov
 * April, 1997
 *
 * Ratpoison windowmanager specific hacking; removed a lot of junk
 * and added keyboard functionality
 * Jonathan Walther
 * krooger@debian.org
 * September, 2001
 *
 * Provided initial implementation of -persist, inspiring
 * the version that actually made it into the code.
 * Rupert Levene 
 * r.levene@lancaster.ac.uk
 * October, 2002
 *
 * Provided a menu-method file to allow ratmenu to access the Debian
 * menu hierarchy.
 * Fabricio Matheus Goncalves
 * fmatheus@bigbross.com
 * April, 2003
 *
 * Provided Xresource handling code
 * Zrajm C Akfohg
 * zrajm@klingonska.org
 * March 2003
 */

#include <stdio.h>
#include <fcntl.h>
#include <signal.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <strings.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <X11/keysym.h>
#include <X11/X.h>
#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/Xatom.h>
#include <X11/Xresource.h>
#include <X11/Xlocale.h>

char version[] = "@(#) ratmenu version 2.3";

#define FONT "9x15bold"
#define	MenuMask (ExposureMask|StructureNotifyMask|KeyPressMask)
#define CTL(c) (c - '`')
#define XTextWidth XmbTextEscapement
#define XLoadQueryFont XLoadQueryFontSet
XFontSet XLoadQueryFontSet(Display *, const char *);

Display *dpy;		/* lovely X stuff */
int screen;
Window root;
Window menuwin;
GC gc;
unsigned long fg;
unsigned long bg;
char *fgcname = NULL;
char *bgcname = NULL;
Colormap dcmap;
XColor color;
XFontSet font;
short font__ascent, font__descent;
Atom wm_protocols;
Atom wm_delete_window;
int g_argc;			/* for XSetWMProperties to use */
char **g_argv;
int savex, savey;
int dpyheight;
Window savewindow;


char *progname;		/* my name */
char *displayname;	/* X display */
char *fontname = NULL;  /* font */
char *labelname;	/* window and icon name */

char **labels;		/* list of labels and commands */
char **commands;
int numitems;
int startitem;          /* which item to highlight first */
int curitem;
int olditem;
int visible;            /* number of items visible on screen */
int off;                /* offset used by dreary */
enum {unset=0, left=1, yes=1, snazzy=1, center=2, no=2, dreary=2, right=3} align, persist, menustyle;

char *shell = "/bin/sh";	/* default shell */
char *prevmenu = NULL;          /* previous menu */

void ask_wm_for_delete(void);
void reap(int);
void redraw_snazzy(int, int, int, int);
void redraw_dreary(int, int, int, int);
void (*redraw) (int, int, int, int);
void run_menu();
void set_wm_hints(int, int);
void spawn(char*);
void usage(void);
void xresources(void);
void defaults(void);

XFontSet XLoadQueryFontSet(Display *disp, const char *fontset_name)
{
  XFontSet fontset;
  int  missing_charset_count;
  char **missing_charset_list;
  char *def_string;
  
  fontset = XCreateFontSet(disp, fontset_name,
                           &missing_charset_list, &missing_charset_count,
                           &def_string);
  if (missing_charset_count /* && debug */) {
    fprintf(stderr, "Missing charsets in FontSet(%s) creation.\n", fontset_name);
    XFreeStringList(missing_charset_list);
  }
  return fontset;
}


/* main --- crack arguments, set up X stuff, run the main menu loop */

int
main(int argc, char **argv)
{
	int i, j;
	char *cp;
	XGCValues gv;
	unsigned long mask;
	XFontSetExtents *extent;

	g_argc = argc;
	g_argv = argv;

	setlocale(LC_CTYPE, "");

	/* set default label name */
	if ((cp = strrchr(argv[0], '/')) == NULL)
		labelname = argv[0];
	else
		labelname = ++cp;

	/* and program name for diagnostics */
	progname = labelname;

	for (i = 1; i < argc; i++) {
		if (strcmp(argv[i], "-display") == 0) {
			displayname = argv[++i];
                } else if (strcmp(argv[i], "-font") == 0) {
                        fontname = argv[++i];
		} else if (strcmp(argv[i], "-label") == 0) {
			labelname = argv[++i];
		} else if (strcmp(argv[i], "-shell") == 0) {
			shell = argv[++i];
		} else if (strcmp(argv[i], "-back") == 0) {
			prevmenu = argv[++i];
		} else if (strcmp(argv[i], "-fg") == 0) {
			fgcname = argv[++i];
		} else if (strcmp(argv[i], "-bg") == 0) {
			bgcname = argv[++i];
		} else if (strcmp(argv[i], "-io") == 0) {
			startitem = atoi(argv[++i]) - 1;
		} else if (strcmp(argv[i], "-persist") == 0) {
			persist = yes;
		} else if (strcmp(argv[i], "-align") == 0) {
			if (strcmp(argv[i+1], "left") == 0) {
				align = left;
			} else if (strcmp(argv[i+1], "center") == 0) {
				align = center;
			} else if (strcmp(argv[i+1], "right") == 0) {
				align = right;
			} else {
				usage();
			}
			i++;
		} else if (strcmp(argv[i], "-style") == 0) {
			if (strcmp(argv[i+1], "snazzy") == 0) {
				menustyle = snazzy;
			} else if (strcmp(argv[i+1], "dreary") == 0) {
				menustyle = dreary;
			} else {
				usage();
			}
			i++;
		} else if (strcmp(argv[i], "-version") == 0) {
			printf("%s\n", version);
			exit(0);
		} else if (argv[i][0] == '-') {
			usage();
		} else {
			break;
		}
	}

	if ((argc-i) % 2 == 1) /* every menu item needs a label AND command */
		usage();

	if ((argc-i) * 2 <= 0)
		usage();

	numitems = (argc-i) / 2;

	labels   = (char **) malloc(numitems * sizeof(char *));
	commands = (char **) malloc(numitems * sizeof(char *));

	if (commands == NULL || labels == NULL) {
		fprintf(stderr, "%s: no memory!\n", progname);
		exit(1);
	}

	for (j=0; i < argc; j++) {
		labels[j]   = argv[i++];
		commands[j] = argv[i++];
	}

	dpy = XOpenDisplay(displayname);
	if (dpy == NULL) {
		fprintf(stderr, "%s: cannot open display", progname);
		if (displayname != NULL)
			fprintf(stderr, " %s", displayname);
		fprintf(stderr, "\n");
		exit(1);
	}
	screen = DefaultScreen(dpy);
	dcmap = DefaultColormap (dpy, screen);

	xresources();
	defaults();

	if ((font = XLoadQueryFont(dpy, fontname)) == NULL) {
		fprintf(stderr, "%s: fatal: cannot load font %s\n", progname, fontname);
		exit(1);
	}
	extent = XExtentsOfFontSet(font);
	font__ascent = extent->max_logical_extent.height * 4 / 5;
	font__descent = extent->max_logical_extent.height / 5;

	gv.foreground = fg^bg;
	gv.background = bg;
	gv.function = GXxor;
	gv.line_width = 0;
	mask = GCForeground | GCBackground | GCFunction | GCLineWidth;

	root = RootWindow(dpy, screen);
	gc = XCreateGC(dpy, root, mask, &gv);

	dpyheight = DisplayHeight(dpy, screen);

	signal(SIGCHLD, reap);

	run_menu();

	XCloseDisplay(dpy);
	exit(0);
}

/* spawn --- run a command */

void
spawn(char *com)
{
	char *sh_basename;

	if ((sh_basename = strrchr(shell, '/')) != NULL) {
		sh_basename++;
	} else {
		sh_basename = shell;
	}

	XCloseDisplay(dpy);
	execl(shell, sh_basename, "-c", com, NULL);
	execl("/bin/sh", "sh", "-c", com, NULL);
	exit(1);
}

/* reap --- collect dead children */

void
reap(int s)
{
	(void) wait((int *) NULL);
	signal(s, reap);
}

/* usage --- print a usage message and die */

void
usage(void)
{
	fprintf(stderr, "usage: %s [-display displayname] [-style {snazzy|dreary}]\n"
			"[-shell shell] [-label name] [-align {left|center|right}]\n"
			"[-fg fgcolor] [-bg bgcolor] [-font fname] [-back prevmenu]\n"
			"[-persist] [-version] [-io offset] [menuitem command] ...\n", progname);
	exit(0);
}

/* run_menu --- put up the window, execute selected commands */

void
run_menu(void)
{
	KeySym key;
	XEvent ev;
	XClientMessageEvent *cmsg;
	int i, wide, high, dx, dy;

	for (i = 0, dx = 0, wide = 0; i < numitems; i++) {
		wide = XTextWidth(font, labels[i], strlen(labels[i])) + 4;
		if (wide > dx)
			dx = wide;
	}
	wide = dx;

	high = font__ascent + font__descent + 1;
	visible = dpyheight / high;
	if (visible > numitems) { visible = numitems; }
	dy = visible * high;

	set_wm_hints(dx, dy);

	ask_wm_for_delete();


	XSelectInput(dpy, menuwin, MenuMask);

	XMapWindow(dpy, menuwin);

	if (startitem < 0) {
		curitem = 0;
	} else if (startitem >= numitems) {
		curitem = numitems - 1;
	} else {
		curitem = startitem;
	}

	if (menustyle == snazzy) {
		redraw = redraw_snazzy;
	} else if (menustyle == dreary)  {
		redraw = redraw_dreary;
		olditem = curitem;
	}

	off = 0;

	for (;;) {
		char keystr[10];
		int keystr_len;
		XNextEvent(dpy, &ev);
		switch (ev.type) {
		case KeyPress:
			keystr_len = XLookupString(&ev.xkey, keystr, sizeof(keystr), &key, NULL);
			if (keystr_len == 1) {
				/* construct some of the "easier" keybindings
				   by looking at the string representation of
				   what got hit, and faking one of the KeySyms
				   that we already handle.
				*/
				switch (keystr[0]) {
				case 'q':
				case '\033':
				case CTL('g'):
					key = XK_Escape;
					break;
				case CTL('p'):
					key = XK_Up;
					break;
				case CTL('n'):
				case CTL('i'): /* c-i is *like* tab */
					key = XK_Down;
					break;
				case CTL('m'):
				case CTL('j'):
				case CTL('f'):
					key = XK_Right;
					break;
				case CTL('b'):
					key = XK_Left;
					break;
				}
			}

			switch (key) {
			case XK_Escape: case XK_q:
				return;
				break;
			case XK_Left: case XK_h:
				if (prevmenu) {
					spawn(prevmenu);
				}
				break;
			case XK_Right: case XK_Return: case XK_l:
				spawn(commands[curitem]);
				break;
			case XK_Tab: case XK_space: case XK_Down: case XK_j: case XK_plus:
				++curitem >= numitems ? curitem =            0 : 0 ;
				redraw(curitem, high, wide, 0);
				break;
			case XK_BackSpace: case XK_Up: case XK_k: case XK_minus:
				curitem-- <= 0        ? curitem = numitems - 1 : 0 ;
				redraw(curitem, high, wide, 0);
				break;
			}
			break;
		case MapNotify: case Expose: redraw(curitem, high, wide, 1); break;
		case UnmapNotify: if (persist == no) return; else XClearWindow(dpy, menuwin); break;
		case CirculateNotify: if (persist == no) return; break;
		case ClientMessage:
			cmsg = &ev.xclient;
			if (cmsg->message_type == wm_protocols
				&& cmsg->data.l[0] == wm_delete_window)
				return;
		}
	}
}

/* set_wm_hints --- set all the window manager hints */

void
set_wm_hints(int wide, int high)
{
	XWMHints *wmhints;
	XSizeHints *sizehints;
	XClassHint *classhints;
	XTextProperty wname;

	if ((sizehints = XAllocSizeHints()) == NULL) {
		fprintf(stderr, "%s: could not allocate size hints\n",
			progname);
		exit(1);
	}

	if ((wmhints = XAllocWMHints()) == NULL) {
		fprintf(stderr, "%s: could not allocate window manager hints\n",
			progname);
		exit(1);
	}

	if ((classhints = XAllocClassHint()) == NULL) {
		fprintf(stderr, "%s: could not allocate class hints\n",
			progname);
		exit(1);
	}

	/* fill in hints in order to parse geometry spec */
	sizehints->width = sizehints->min_width = sizehints->max_width = wide;
	sizehints->height = sizehints->min_height = sizehints->max_height = high;
	sizehints->flags = USSize|PSize|PMinSize|PMaxSize;

	if (XStringListToTextProperty(& labelname, 1, & wname) == 0) {
		fprintf(stderr, "%s: could not allocate window name structure\n",
			progname);
		exit(1);
	}

	menuwin = XCreateSimpleWindow(dpy, root, sizehints->x, sizehints->y,
				sizehints->width, sizehints->height, 1, fg, bg);

	wmhints->input = True;
	wmhints->initial_state = NormalState;
	wmhints->flags = StateHint | InputHint;

	classhints->res_name = progname;
	classhints->res_class = "ratmenu";

	XSetWMProperties(dpy, menuwin, & wname, NULL,
		g_argv, g_argc, sizehints, wmhints, classhints);
}

/* ask_wm_for_delete --- jump through hoops to ask WM to delete us */

void
ask_wm_for_delete(void)
{
	int status;

	wm_protocols = XInternAtom(dpy, "WM_PROTOCOLS", False);
	wm_delete_window = XInternAtom(dpy, "WM_DELETE_WINDOW", False);
	status = XSetWMProtocols(dpy, menuwin, & wm_delete_window, 1);

	if (status != True)
		fprintf(stderr, "%s: could not ask for clean delete\n",
			progname);
}

/* redraw --- actually draw the menu */

void
redraw_snazzy (int curitem, int high, int wide, int fullredraw)
{
	int i, j, ty, tx;

	XClearWindow(dpy, menuwin);
	for (i = 0, j = curitem; i < visible; i++, j++) {
		j %= numitems;
		if (align == left) {
			tx = 0;
		} else if (align == center) {
                	tx = (wide - XTextWidth(font, labels[j], strlen(labels[j]))) / 2;
		} else {/* align == right */
                	tx = wide - XTextWidth(font, labels[j], strlen(labels[j]));
		}
                ty = i*high + font__ascent + 1;
                XmbDrawString(dpy, menuwin, font, gc, tx, ty, labels[j], strlen(labels[j]));
	}
	XFillRectangle(dpy, menuwin, gc, 0, 0, wide, high); 
}

void
redraw_dreary (int curitem, int high, int wide, int fullredraw)
{
	int i, ty, tx, oldoff;

	oldoff = off;
	if (curitem + 1 > visible + off) {
		off = curitem + 1 - visible;
	} else if (curitem + 1 <= off) {
		off = curitem;
	}
	if (fullredraw || oldoff != off) {
		XClearWindow(dpy, menuwin);
		for (i = 0; i < visible; i++) {
			if (align == left) {
				tx = 0;
			} else if (align == center) {
                		tx = (wide - XTextWidth(font, labels[i+off], strlen(labels[i+off]))) / 2;
			} else {/* align == right */
                		tx = wide - XTextWidth(font, labels[i+off], strlen(labels[i+off]));
			}
                	ty = i*high + font__ascent + 1;
                	XmbDrawString(dpy, menuwin, font, gc, tx, ty, labels[i+off], strlen(labels[i+off]));
		}
		XFillRectangle(dpy, menuwin, gc, 0, (curitem-off)*high, wide, high); 
	} else {
		XFillRectangle(dpy, menuwin, gc, 0, (olditem-off)*high, wide, high); 
		XFillRectangle(dpy, menuwin, gc, 0, (curitem-off)*high, wide, high); 
	}
	olditem = curitem;
}

void
xresources(void)
{
	char *res;
	
	if (bgcname == NULL) {
		if ((res = XGetDefault(dpy, progname, "bgcolor")) != NULL ) {
			bgcname = strdup(res);
		}
	}
	if (fgcname == NULL) {
		if ((res = XGetDefault(dpy, progname, "fgcolor")) != NULL ) {
			fgcname = strdup(res);
		}
	}
	if (fontname == NULL) {
		if ((res = XGetDefault(dpy, progname, "font")) != NULL ) {
			fontname = strdup(res);
		}
	}
	if (menustyle == unset) {
		if ((res = XGetDefault(dpy, progname, "style")) != NULL ) {
			if      (strcasecmp(res, "dreary") == 0) menustyle = dreary;
			else if (strcasecmp(res, "snazzy") == 0) menustyle = snazzy;
		}
	}
	if (persist == unset) {
		if ((res = XGetDefault(dpy, progname, "persist")) != NULL ) {
			if      (strcasecmp(res, "on"  )  == 0) persist = yes;
			else if (strcasecmp(res, "off")   == 0) persist = no;
			else if (strcasecmp(res, "yes")   == 0) persist = yes;
			else if (strcasecmp(res, "no")    == 0) persist = no;
			else if (strcasecmp(res, "true")  == 0) persist = yes;
			else if (strcasecmp(res, "false") == 0) persist = no;
		}
	}
	if (align == unset) {
		if ((res = XGetDefault(dpy, progname, "align")) != NULL ) {
			if      (strcasecmp(res, "left"  ) == 0) align = left;
			else if (strcasecmp(res, "center") == 0) align = center;
			else if (strcasecmp(res, "right" ) == 0) align = right;
		}
	}
}

void
defaults(void)
{
	if (bgcname == NULL || XParseColor(dpy, dcmap, bgcname, &color) == 0 || XAllocColor(dpy, dcmap, &color) == 0) {
		bg = BlackPixel(dpy, screen);
	} else { bg = color.pixel; }
	if (fgcname == NULL || XParseColor(dpy, dcmap, fgcname, &color) == 0 || XAllocColor(dpy, dcmap, &color) == 0) {
		fg = WhitePixel(dpy, screen);
	} else { fg = color.pixel; }
	if (fontname == NULL) { fontname = FONT; }
	if (persist == unset) persist = no;
	if (align == unset) align = left;
	if (menustyle == unset) menustyle = snazzy;
}
